pkg/sync/syncer.go

```diff
@@ -745,22 +814,91 @@ func (s *syncer) syncGrantsForResource(ctx context.Context, resourceID *v2.Resou
 	if err != nil {
 		return err
 	}
-	for _, grant := range resp.List {
+
+	nextPageToken := resp.NextPageToken
+
+	respAnnos := annotations.Annotations(resp.GetAnnotations())
+	etagMatch := respAnnos.Contains(&v2.ETagMatch{})
+
+	var grantsRet []*v2.Grant
+
+	// We have a previous etag, and the connector has indicated an etag match
+	if etagMatch {
+		if prevEtag == nil {
+			return errors.New("connector returned an etag match - but there is no previous sync generation to use")
+		}
+
+		// We have a previous sync, and the connector would like to use the previous sync results
+		var npt string
+		// Fetch the grants for this resource from the previous sync, and store them in the current sync.
+		storeAnnos := annotations.Annotations{}
+		storeAnnos.Update(&c1zpb.SyncDetails{
+			Id: prevSyncID,
+		})
+
+		for {
+			prevGrantsResp, err := s.store.ListGrants(ctx, &v2.GrantsServiceListGrantsRequest{
+				Resource:    resource,
+				Annotations: storeAnnos,
+				PageToken:   npt,
+				PageSize:    1000,
+			})
+			if err != nil {
+				return err
+			}
+
+			grantsRet = append(grantsRet, prevGrantsResp.List...)
+
+			if prevGrantsResp.NextPageToken == "" {
+				break
+			}
+			npt = prevGrantsResp.NextPageToken
+		}
+	}
+
+	// We want to process any grants from the previous sync first so that if there is a conflict
+	grantsRet = append(grantsRet, resp.List...)
+
+	for _, grant := range grantsRet {
 		err = s.store.PutGrant(ctx, grant)
 		if err != nil {
 			return err
 		}
 	}
 
-	s.handleProgress(ctx, s.state.Current(), len(resp.List))
+	s.handleProgress(ctx, s.state.Current(), len(grantsRet))
 
-	if resp.NextPageToken != "" {
-		err = s.state.NextPage(ctx, resp.NextPageToken)
+	if nextPageToken != "" {
+		err = s.state.NextPage(ctx, nextPageToken)
 		if err != nil {
 			return err
 		}
+		return nil
+	}
+
+	s.state.FinishAction(ctx)
+
+	var updatedETag *v2.ETag
+	if etagMatch {
+		updatedETag = prevEtag
 	} else {
-		s.state.FinishAction(ctx)
+		newETag := &v2.ETag{}
+		ok, err := respAnnos.Pick(newETag)
+		if err != nil {
+			return err
+		}
+		if ok {
+			updatedETag = newETag
+		}
+	}
+
+	if updatedETag != nil {
+		resourceAnnos.Update(updatedETag)
```

Comment:
Oh so any grants call updates the resources etag, so they could be overwritten by a next page.

Reply:
We only reach this portion of the code if we're on the last page. If you look at line 871, if there is a next page token we exit early.